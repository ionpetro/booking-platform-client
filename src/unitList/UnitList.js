import Drawer from '../components/Drawer/Drawer.vue';
import Message from '../components/Message/Message.vue';
import UnitCard from './unitCard/UnitCard.vue';
import UnitDetails from './unitDetails/UnitDetails.vue';
import UnitService from './unit.service';

const unitService = new UnitService();

export default {
  name: 'UnitList',
  components: {
    UnitDetails,
    Drawer,
    UnitCard,
    'v-message': Message
  },
  data() {
    return {
      message: '',
      loading: false,
      page: 1,
      size: 16,
      totalPages: 1,
      units: [],
      isDrawerOpened: false,
      selectedUnitId: undefined
    };
  },
  mounted() {
    this.fetch();
  },
  methods: {
    async fetch() {
      this.loading = true;
      try {
        const response = await unitService.getUnits(this.page, this.size);
        this.units = [...this.units, ...response.units];
        this.totalPages = response.totalPages;
      } catch (error) {
        this.message = error.message;
      } finally {
        this.loading = false;
      }
    },
    visibilityChanged(isVisible) {
      if (!isVisible || this.page >= this.totalPages) {
        return;
      }
      this.page += 1;
      this.fetch();
    },
    openDrawer(id) {
      this.isDrawerOpened = true;
      this.selectedUnitId = id;
    },
    onDrawerClosed() {
      this.isDrawerOpened = false;
    }
  }
};
