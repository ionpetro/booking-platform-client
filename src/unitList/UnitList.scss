.list {
  margin: auto;
  margin-top: 116px;
  width: 1744px;
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;

  &__error {
    position: fixed;
    z-index: 9999;
    margin: 50px auto;
    left: 0;
    right: 0;
  }

  &__card {
    margin: auto;
    outline: none;
  }

  @media only screen and (max-width: 1919px) {
    width: 1300px;
  }

  @media only screen and (max-width: 1440px) {
    width: 840px;
  }

  @media only screen and (max-width: $medium-breakpoint) {
    width: 680px;
  }

  @media only screen and (max-width: 720px) {
    width: auto;
    align-items: center;
    margin-left: 16px;
    margin-right: 16px;
  }
}

.visibilityElement {
  height: 10px;
}
