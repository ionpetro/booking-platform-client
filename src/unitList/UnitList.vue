<template>
  <div>
    <transition name="fade">
      <v-message class="list__error" :type="'error'" v-if="message">{{ message }}</v-message>
    </transition>
    <Drawer v-if="isDrawerOpened" @onDrawerClosed="onDrawerClosed"
      ><UnitDetails :selectedUnitId="selectedUnitId"></UnitDetails
    ></Drawer>
    <div class="list">
      <div
        v-for="unit of units"
        :key="unit.id"
        class="list__card"
        tabindex="0"
        @click="openDrawer(unit.id)"
      >
        <UnitCard :unit="unit"></UnitCard>
      </div>
    </div>
    <div
      class="visibilityElement"
      v-if="units.length"
      v-observe-visibility="{
        callback: visibilityChanged,
        throttle: 200
      }"
    ></div>
  </div>
</template>

<script src="./UnitList.js"></script>

<style lang="scss" src="./UnitList.scss" scoped></style>
