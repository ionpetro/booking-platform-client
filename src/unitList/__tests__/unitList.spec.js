import {shallowMount} from '@vue/test-utils';
import flushPromises from 'flush-promises';
import UnitList from '../UnitList.vue';
import UnitService from '../unit.service';
import UnitMock from '../unitDetails/__tests__/mocks/unit.mock';

jest.mock('../unit.service', () => jest.fn());

const mockData = [UnitMock];
const mockResp = {units: mockData, totalPages: 4, itemsCount: 48};

describe('UnitList', () => {
  let wrapper;

  beforeEach(() => {
    UnitService.mockClear();
  });

  afterEach(() => {
    wrapper.destroy();
  });

  describe('when loaded', () => {
    it('renders correctly', () => {
      wrapper = shallowMount(UnitList);
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('when component is mounted', () => {
    it('fetches data correctly', async () => {
      UnitService.prototype.getUnits = jest.fn(() => Promise.resolve(mockResp));

      wrapper = shallowMount(UnitList);

      await flushPromises();

      expect(wrapper.vm.$data.totalPages).toEqual(4);
      expect(wrapper.vm.$data.units).toEqual(mockData);
      expect(wrapper.vm.$data.loading).toBeFalsy();
      expect(wrapper.vm.$data.message).toEqual('');
    });

    it('handles fetch data error', async () => {
      UnitService.prototype.getUnits = jest.fn(() => Promise.reject(new Error('error')));

      wrapper = shallowMount(UnitList);

      await flushPromises();

      expect(wrapper.vm.$data.message).toEqual('error');
      expect(wrapper.vm.$data.totalPages).toEqual(1);
      expect(wrapper.vm.$data.units).toEqual([]);
      expect(wrapper.vm.$data.loading).toBeFalsy();
    });
  });
});
