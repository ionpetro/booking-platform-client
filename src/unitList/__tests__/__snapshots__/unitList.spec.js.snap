// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnitList when loaded renders correctly 1`] = `
<div>
  <transition-stub
    name="fade"
  >
    <!---->
  </transition-stub>
   
  <!---->
   
  <div
    class="list"
  />
   
  <!---->
</div>
`;
